pact-request-api/Pact/Core/Command/Client.hs

```diff
@@ -22,11 +77,682 @@ import Pact.Core.Command.Types
 import Pact.Core.Command.Util
 import Pact.Core.Command.Crypto
 import Pact.Core.Guards
-import qualified Pact.Core.Hash as PactHash
 import Pact.Core.PactValue
 import Pact.Core.Names
 import Pact.Core.Verifiers
+import Pact.Core.StableEncoding
+import Pact.Core.Gas
+import Pact.Core.Hash
+import Pact.Core.SPV
+import qualified Pact.Core.Hash as PactHash
+import Pact.Core.Command.SigData
+
+
+
+-- -------------------------------------------------------------------------- --
+-- ApiKeyPair
+
+-- | For fully-signed commands
+data ApiKeyPair = ApiKeyPair {
+  _akpSecret :: PrivateKeyBS,
+  _akpPublic :: Maybe PublicKeyBS,
+  _akpAddress :: Maybe Text,
+  _akpScheme :: Maybe PPKScheme,
+  _akpCaps :: Maybe [CapToken QualifiedName PactValue]
+  } deriving (Eq, Show, Generic)
+
+instance JD.FromJSON ApiKeyPair where
+  parseJSON = JD.withObject "ApiKeyPair" $ \o -> do
+    secret <- o JD..: "secret"
+    pub <- o JD..:? "public"
+    addr <- o JD..: "address"
+    scheme <- o JD..:? "scheme"
+    caps <- ((fmap.fmap) _stableEncoding <$> o JD..:? "caps")
+    pure $ ApiKeyPair
+      {_akpSecret = secret
+      , _akpPublic = pub
+      , _akpAddress = addr
+      , _akpScheme = scheme
+      , _akpCaps = caps}
+
+
+instance J.Encode ApiKeyPair where
+  build o = J.object
+    [ "address" J..= _akpAddress o
+    , "secret" J..= _akpSecret o
+    , "scheme" J..= _akpScheme o
+    , "caps" J..= fmap (J.Array . fmap StableEncoding) (_akpCaps o)
+    , "public" J..= _akpPublic o
+    ]
+  {-# INLINE build #-}
+
+-- -------------------------------------------------------------------------- --
+-- ApiSigner
+
+-- | For unsigned commands
+data ApiSigner = ApiSigner {
+  _asPublic :: Text,
+  _asAddress :: Maybe Text,
+  _asScheme :: Maybe PPKScheme,
+  _asCaps :: Maybe [CapToken QualifiedName PactValue]
+  } deriving (Eq, Show, Generic)
+
+instance JD.FromJSON ApiSigner where
+  parseJSON = JD.withObject "ApiSigner" $ \o -> do
+    pub <- o JD..: "public"
+    addr <- o JD..:? "address"
+    scheme <- o JD..:? "scheme"
+    caps <- ((fmap.fmap) _stableEncoding <$> o JD..:? "caps")
+    pure $ ApiSigner
+      { _asPublic = pub
+      , _asAddress = addr
+      , _asScheme = scheme
+      , _asCaps = caps}
+
+
+instance J.Encode ApiSigner where
+  build o = J.object
+    [ "address" J..= _asAddress o
+    , "scheme" J..= _asScheme o
+    , "caps" J..= fmap (J.Array . fmap StableEncoding) (_asCaps o)
+    , "public" J..= _asPublic o
+    ]
+  {-# INLINE build #-}
+
+-- -------------------------------------------------------------------------- --
+-- ApiPublicMeta
+
+data ApiPublicMeta = ApiPublicMeta
+  { _apmChainId :: Maybe ChainId
+  , _apmSender :: Maybe Text
+  , _apmGasLimit :: Maybe GasLimit
+  , _apmGasPrice :: Maybe GasPrice
+  , _apmTTL :: Maybe TTLSeconds
+  , _apmCreationTime :: Maybe TxCreationTime
+  } deriving (Eq, Show, Generic)
+
+(.:@?) :: JD.FromJSON (StableEncoding b) => JD.Object -> JD.Key -> A.Parser (Maybe b)
```

Comment:
😭 ... ok

Reply:
hahaha we can remove this, It just saved me some lines of code.